### ISSUE TITLE
Use of `ExtensionPoint` in TasksApplication is confusing

### ISSUE THREAD

**Issue by mdickinson** [enthought/envisage#102]
[Arising from conversations with @stefanoborini]

There's a surprising and potentially confusing use of the `ExtensionPoint` trait type in the `TasksApplication`: the extension points in the application appear to duplicate those in the `TasksPlugin`. In the `TasksApplication` class definition, we have:

```python
class TasksApplication(Application):

    ...

    # Extension point IDs.
    TASK_FACTORIES = 'envisage.ui.tasks.tasks'
    TASK_EXTENSIONS = 'envisage.ui.tasks.task_extensions'

    ...

    # Contributed task factories. This attribute is primarily for run-time
    # inspection; to instantiate a task, use the 'create_task' method.
    task_factories = ExtensionPoint(id=TASK_FACTORIES)

    # Contributed task extensions.
    task_extensions = ExtensionPoint(id=TASK_EXTENSIONS)
```

In the `TasksPlugin` definition, we have (where `PKG` evaluates to `envisage.ui.tasks`)

```python
class TasksPlugin(Plugin):

    ...

    TASKS = PKG + '.tasks'
    TASK_EXTENSIONS = PKG + '.task_extensions'

    ...

    tasks = ExtensionPoint(
        List(Instance('envisage.ui.tasks.task_factory.TaskFactory')),
        id=TASKS,
        desc="""

        This extension point makes tasks avaiable to the application.

        Each contribution to the extension point must be an instance of
        'envisage.tasks.api.TaskFactory.
        """)

    task_extensions = ExtensionPoint(
        List(Instance('envisage.ui.tasks.task_extension.TaskExtension')),
        id=TASK_EXTENSIONS,
        desc="""

        This extension point permits the contribution of new actions and panes
        to existing tasks (without creating a new task).

        Each contribution to the extension point must be an instance of
        'envisage.tasks.api.TaskExtension'.
        """)
```

What seems to be happening here is that only the `TasksPlugin` definitions are relevant for the purposes of creating and registering the extension points: the `ExtensionPoint.connect_extension_point_traits` method only gets called for plugin objects, not for the application object. The `TasksApplication` redeclarations then provide a lazy way of accessing the list of current task factories / task extensions. In particular, this means that attempting to listen to `task_factories_items` (for example) on the application won't work, since the relevant trait connections haven't been hooked up.

At the least, I'd suggest documenting these uses in `TasksApplication` a little better to indicate what's going on. (There's a comment for the `task_factories` trait that's suggestive, but isn't explicit that this isn't really where the extension point is being defined.) But we may also be able to find a better way to do this.

**Comment by stefanoborini** [comment]
The immediate consequence of this is that if one declares the extension points on the application only, they will be empty.

**Comment by mdickinson** [comment]
> The immediate consequence of this is that if one declares the extension points on the application only, they will be empty.

Right, and it's surprising and a bit disturbing that we don't simply get an error when looking up extension points for an extension id that hasn't been registered.

https://github.com/enthought/envisage/blob/4.6.0/envisage/extension_registry.py#L160-L163
